### PR TITLE
Add BinaryBuffer for data serialization; make DataPacket serializable

### PR DESCRIPTION
**Pull request by skylian**

### CODE REVIEW

File: data_packet.cpp [hunk 1]
```diff
@@ -118,4 +120,65 @@ size_t StateBuffer::get_id_size() {
...
+        str_ && other.str_ && *str_ != *other.str_) {
+        return false;
+    }
+
+    return true;
 }
+
+void StateBuffer::encode(util::BinaryBuffer& buf) {
+    uint8_t flags = (reals_ ? 1 : 0) | (pixels_ ? 2 : 0) |
+                    (id_ ? 4 : 0) | (str_ ? 8 : 0);
```
Comment (Haichao-Zhang) [inline data_packet.cpp]: It might be better to use a enum type for the flags (1, 2, 4 ,8) both here and in line 166-178?

File: memory_util.h [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+}
+
+inline BinaryBuffer::BinaryBuffer(size_type size) {
+    data_ = (size != 0 ? new uchar[size] : NULL);
+    size_ = capacity_ = size;
+    own_ = true;
+}
+
+inline BinaryBuffer::BinaryBuffer(size_type size, size_type capacity) {
+    if (size > capacity) {
```
Comment (Haichao-Zhang) [inline memory_util.h]: What's the difference and relationship between size and capacity? Adding a brief explanation to each parameter would be better.

File: .gitignore [hunk 3]
```diff
@@ -17,3 +17,4 @@ build
 examples/test_atari
 examples/test_xworld
 examples/test_simple_race
+examples/test_xworld3d
```
Comment (yu239) [inline .gitignore]: This should be moved to the previous commit.

File: data_packet.cpp [hunk 4]
```diff
@@ -118,4 +120,65 @@ size_t StateBuffer::get_id_size() {
         return 0;
     }
 }
+
+bool StateBuffer::operator==(const StateBuffer& other) {
+    if (reals_ && !other.reals_ || !reals_ && other.reals_ ||
```
Comment (yu239) [inline data_packet.cpp]: Write a function for comparing two pointers

File: data_packet.cpp [hunk 5]
```diff
@@ -118,4 +120,65 @@ size_t StateBuffer::get_id_size() {
...
+    }
+    if (id_) {
+        buf.append(*id_);
+    }
+    if (str_) {
+        buf.append(*str_);
+    }
+}
+
+void StateBuffer::decode(util::BinaryBuffer& buf) {
```
Comment (yu239) [inline data_packet.cpp]: const util::BinaryBuffer& buf

File: data_packet.cpp [hunk 6]
```diff
@@ -118,4 +120,65 @@ size_t StateBuffer::get_id_size() {
...
+    uint8_t flags = 0;
+    buf.read(flags);
+    if (flags & 1) {
+        reals_ = std::make_shared<std::vector<float>>();
```
Comment (yu239) [inline data_packet.cpp]: reals_->clear();

File: data_packet.cpp [hunk 7]
```diff
@@ -118,4 +120,65 @@ size_t StateBuffer::get_id_size() {
...
+        buf.read(*reals_);
+    }
+    if (flags & 2) {
+        pixels_ = std::make_shared<std::vector<uint8_t>>();
```
Comment (yu239) [inline data_packet.cpp]: pixels_->clear();

File: data_packet.cpp [hunk 8]
```diff
@@ -118,4 +120,65 @@ size_t StateBuffer::get_id_size() {
...
+        buf.read(*pixels_);
+    }
+    if (flags & 4) {
+        id_ = std::make_shared<std::vector<int>>();
```
Comment (yu239) [inline data_packet.cpp]: ids_->clear();

File: data_packet.cpp [hunk 9]
```diff
@@ -118,4 +120,65 @@ size_t StateBuffer::get_id_size() {
...
+        buf.read(*id_);
+    }
+    if (flags & 8) {
+        str_ = std::make_shared<std::string>();
```
Comment (yu239) [inline data_packet.cpp]: str_->clear();

File: data_packet.h [hunk 10]
```diff
@@ -300,7 +306,27 @@ class DataPacket {
         }
     }
 
-  private:
+    void encode(util::BinaryBuffer& buf) {
```
Comment (yu239) [inline data_packet.h]: Move definitions to CPP files. Only keep short ones in header files.

File: memory_util.h [hunk 11]
```diff
@@ -0,0 +1,298 @@
+#pragma once
```
Comment (yu239) [inline memory_util.h]: Add file header copyright

File: memory_util.h [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+#include <cstring>
+#include <stdexcept>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+namespace simulator {
+namespace util {
+
+typedef std::size_t size_type;
```
Comment (yu239) [inline memory_util.h]: Why such typedef? You can just use size_t without std::

File: memory_util.h [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+
+inline BinaryBuffer& BinaryBuffer::operator=(const BinaryBuffer& other) {
+    if (&other != this) {
+        if (other.size_ > capacity_) {
+            delete_if_own();
+            data_ = new uchar[other.capacity_];
+            capacity_ = other.capacity_;
+        }
+        std::memcpy(data_, other.data_, other.size_);
+        size_ = other.size_;
```
Comment (yu239) [inline memory_util.h]: own_ = false?

File: memory_util.h [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+        delete_if_own();
+        data_ = new uchar[size];
+        capacity_ = size;
+        own_ = true;
+    }
+    std::memcpy(data_, data, size);
+    size_ = size;
+}
+
+inline void BinaryBuffer::take_over(void* data, size_type size) {
```
Comment (yu239) [inline memory_util.h]: How can this let the outside know that data is taken over? What if the outside code tries to delete data?

File: memory_util.h [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+
+template <typename T>
+inline void BinaryBuffer::read(T& t) {
+    read(&t, 1);
+}
+
+inline void BinaryBuffer::read(std::string& str) {
+    std::size_t len;
+    read(len);
+    char* tmp = new char[len+1];
```
Comment (yu239) [inline memory_util.h]: You can just use stack-based mem allocation.
"char tmp[len+1];"
Then you don't have to delete [] tmp.

File: memory_util.h [hunk 16]
```diff
@@ -0,0 +1,298 @@
+#pragma once
+
+#include <cassert>
```
Comment (yu239) [inline memory_util.h]: Use CHECK instead.

File: tests/test_binary_buffer.cpp [hunk 17]
```diff
@@ -0,0 +1,221 @@
+/**
```
Comment (yu239) [inline tests/test_binary_buffer.cpp]: change header to copyright

File: tests/test_statepacket.cpp [hunk 18]
```diff
@@ -0,0 +1,105 @@
+/**
```
Comment (yu239) [inline tests/test_statepacket.cpp]: change header to copyright

File: memory_util.h [hunk 19]
```diff
@@ -0,0 +1,298 @@
...
+#include <cstring>
+#include <stdexcept>
+#include <string>
+#include <type_traits>
+#include <vector>
+
```
Comment (yu239) [inline memory_util.h]: Do we need all the functions in this file? It seems a little bit overkill. Try to keep the minimal version.

**Pull request merged**